LAPACKE/src/lapacke_clarfb.c

```diff
@@ -42,67 +42,37 @@ lapack_int LAPACKE_clarfb( int matrix_layout, char side, char trans, char direct
     lapack_int info = 0;
     lapack_int ldwork;
     lapack_complex_float* work = NULL;
-    lapack_int ncols_v, nrows_v;
+    lapack_int nrows_v, ncols_v;
+    lapack_logical left, col, forward;
+    char uplo;
     if( matrix_layout != LAPACK_COL_MAJOR && matrix_layout != LAPACK_ROW_MAJOR ) {
         LAPACKE_xerbla( "LAPACKE_clarfb", -1 );
         return -1;
     }
 #ifndef LAPACK_DISABLE_NAN_CHECK
     if( LAPACKE_get_nancheck() ) {
         /* Optionally check input matrices for NaNs */
-        lapack_int lrv, lcv;  /* row, column stride */
-        if( matrix_layout == LAPACK_COL_MAJOR ) {
-            lrv = 1;
-            lcv = ldv;
-        } else {
-            lrv = ldv;
-            lcv = 1;
-        }
-        ncols_v =     LAPACKE_lsame( storev, 'c' ) ? k :
-                  ( ( LAPACKE_lsame( storev, 'r' ) && LAPACKE_lsame( side, 'l' ) ) ? m :
-                  ( ( LAPACKE_lsame( storev, 'r' ) && LAPACKE_lsame( side, 'r' ) ) ? n : 1) );
+        left = LAPACKE_lsame( side, 'l' );
+        col = LAPACKE_lsame( storev, 'c' );
+        forward = LAPACKE_lsame( direct, 'f' );
 
-        nrows_v =   ( LAPACKE_lsame( storev, 'c' ) && LAPACKE_lsame( side, 'l' ) ) ? m :
-                  ( ( LAPACKE_lsame( storev, 'c' ) && LAPACKE_lsame( side, 'r' ) ) ? n :
-                    ( LAPACKE_lsame( storev, 'r' ) ? k : 1) );
-        if( LAPACKE_cge_nancheck( matrix_layout, m, n, c, ldc ) ) {
-            return -13;
+        nrows_v = ( col && left ) ? m : ( ( col && !left ) ? n : ( !col ? k : 1) );
+        ncols_v = ( !col && left ) ? m : ( ( !col && !left ) ? n : ( col ? k : 1 ) );
+        uplo = ( ( left && col ) || !( left || col ) ) ? 'l' : 'u';
+
+        if( !forward && ( col && k > nrows_v ) || ( !col && k > ncols_v )) {
```

Comment:
Is `if( !forward && ( col && k > nrows_v ) || ( !col && k > ncols_v ))` supposed to be `if(( col && k > nrows_v ) || ( !col && k > ncols_v ))` ? I didn't get the test `!forward`.

In any case, I arrived to this line of code because we should at least add parenthesis to ensure correct behavior. The conditional currently looks like `if( A && B || C )`.

Reply:
I am marking @ACSimon33 and @langou just to make sure you both see my comment.

Reply:
Looking at the code before my PR, there was the same behavior, with a nested if statement:

```
       } else if( LAPACKE_lsame( storev, 'c' ) && LAPACKE_lsame( direct, 'b' ) ) {
            if( k > nrows_v ) {
                LAPACKE_xerbla( "LAPACKE_clarfb", -8 );
                return -8;
            }
            if( LAPACKE_ctr_nancheck( matrix_layout, 'u', 'u', k,
                                      &v[(nrows_v-k)*lrv], ldv ) )
                return -9;
            if( LAPACKE_cge_nancheck( matrix_layout, nrows_v-k, ncols_v, v, ldv ) )
                return -9;
        }
```

The test was there to prevent segfaults when accessing `&v[(nrows_v-k)*lrv]`. These tests were only in the if branches with `direct == b` (`!forward`). So in order to mirror the old behavior we should put parenthesis around the OR statement. The question remains if there is any reason to allow `k > nrows_v` (or `k > ncols_v`) if `direct == f`.

I think we probably can remove the `!forward` test.

Reply:
Related to this is #780. Now I am reading what @ACSimon33 wrote and it is written "n order to mirror the old behavior we should put parenthesis around the OR statement." Shoo. We just did the reverse and put parenthesis around the AND statement.

In any case, the comment "The question remains if there is any reason to allow `k > nrows_v` (or `k > ncols_v`) if `direct == f`. I think we probably can remove the `!forward` test.

We'll work on this and update the thread. The [<T>LAPACK](https://github.com/tlapack/tlapack) team is having a look at larfb these days so hopefully we can also get some input from them.